kyo-combinators/shared/src/main/scala/kyo/constructors.scala

```diff
@@ -0,0 +1,208 @@
+        Resources.ensure(finalizer)
+
+    def async[A](register: (A < Fibers => Unit) => Unit < Fibers)(
+        using Flat[A]
+    ): A < Fibers =
+        for
+            promise <- Fibers.initPromise[A]
+            registerFn = (eff: A < Fibers) =>
+                val effFiber = Fibers.init(eff)
+                val updatePromise =
+                    effFiber.map(_.onComplete(a => promise.complete(a).unit))
+                val updatePromiseIO = Fibers.init(updatePromise).unit
+                IOs.run(updatePromiseIO)
+            _ <- register(registerFn)
+            a <- promise.get
+        yield a
+
+    def attempt[A, S](effect: => A < S)(using Flat[A]): A < (S & Aborts[Throwable]) =
+        Aborts.catching[Throwable](effect)
+
+    def collect[A, S, A1, S1](
+        sequence: => Seq[A] < S
+    )(
+        useElement: PartialFunction[A, A1 < S1]
+    ): Seq[A1] < (S & S1) =
+        sequence.flatMap((seq: Seq[A]) => Seqs.collect(seq.collect(useElement)))
+
+    def debugln[S](message: => String < S): Unit < (S & IOs) =
+        message.map(m => Console.default.println(m))
+
+    def fail[E, S](error: => E < S): Nothing < (S & Aborts[E]) =
+        error.map(e => Aborts.fail(e))
+
+    def foreach[A, S, A1, S1](
+        sequence: => Seq[A] < S
+    )(
+        useElement: A => A1 < S1
+    ): Seq[A1] < (S & S1) =
+        collect(sequence)(a => useElement(a))
+
+    def foreachDiscard[A, S, S1](
+        sequence: => Seq[A] < S
+    )(
+        useElement: A => Any < S1
+    ): Unit < (S & S1) =
+        foreach(sequence)(useElement).unit
+
+    def foreachPar[A, S, A1](
+        sequence: => Seq[A] < S
+    )(
+        useElement: A => A1 < Fibers
+    )(using Flat[A1]): Seq[A1] < (S & Fibers) =
+        sequence.map(seq => Fibers.parallel(seq.map(useElement)))
+
+    def foreachParDiscard[A, S, Any](
+        sequence: => Seq[A] < S
+    )(
+        useElement: A => Any < Fibers
+    )(using Flat[Any]): Unit < (S & Fibers) =
+        sequence.map(seq => Fibers.parallel(seq.map(v => useElement(v).unit))).unit
+
+    def fromAutoCloseable[A <: AutoCloseable, S](closeable: => A < S): A < (S & Resources) =
+        acquireRelease(closeable)(c => IOs(c.close()))
+
+    def fromEither[E, A](either: => Either[E, A]): A < Aborts[E] =
+        Aborts.get(either)
+
+    def fromFuture[A: Flat, S](future: => Future[A] < S): A < (S & Fibers) =
+        future.map(f => Fibers.fromFuture(f))
+
+    def fromPromiseScala[A: Flat, S](promise: => scala.concurrent.Promise[A] < S)
+        : A < (S & Fibers) =
+        promise.map(p => fromFuture(p.future))
+
+    def fromOption[A, S](option: => Option[A] < S): A < (S & Options) =
+        Options.get(option)
+
+    def fromSeq[A, S](sequence: => Seq[A] < S): A < (S & Choices) =
+        sequence.flatMap(seq => Choices.get(seq))
+
+    def fromTry[A, S](_try: => scala.util.Try[A] < S): A < (S & Aborts[Throwable]) =
+        _try.map(t => Aborts.get(t.toEither))
```

Comment:
https://github.com/getkyo/kyo/pull/461

Reply:
Nice, since 461 isn't merged, for now I just improved the implementation to avoid the additional allocation